[16.x] Stripe Basil API Support

Diddyy opened the pull request:
This PR targets the `16.x` branch in line with Laravel’s support policy. It finalizes Cashier’s compatibility with Stripe’s **Basil** API (API version `2025-03-31.basil`), as highlighted in the Cashier source code:

```php
// src/Cashier.php
const STRIPE_VERSION = '2025-03-31.basil';
```
### Summary of Changes
Invoice Line Items: Added helpers to extract pricing data using Stripe’s new `price_details` and `inline_price_data` structures introduced in Basil:

```php
public function priceId()
{
    // Handle the new pricing structure (Basil release)
    if (isset($this->item->pricing) && $this->item->pricing->type === 'price_details') {
        return $this->item->pricing->price_details->price ?? null;
    }

    return null;
}
```
Unit Amount Handling: Retrieve unit amounts and support inline price data:

```php
public function unitAmount()
{
    // Handle the new pricing structure (Basil release)
    if (isset($this->item->pricing)) {
        if (isset($this->item->pricing->unit_amount_decimal)) {
            return (int) $this->item->pricing->unit_amount_decimal;
        }
        if ($this->item->pricing->type === 'inline_price_data' &&
            isset($this->item->pricing->inline_price_data->unit_amount)) {
            return (int) $this->item->pricing->inline_price_data->unit_amount;
        }
    }

    return null;
}
```

Tax Behavior Retrieval: Access the `tax_behavior` on invoice line items:

```php
public function taxBehavior()
{
    $price = $this->price();
    return $price ? ($price->tax_behavior ?? null) : null;
}
```

### Benefits to Users
Keeps Cashier aligned with Stripe’s latest Basil API, as requested by Taylor April tweet inviting contributions.

### Testing
Added PHPUnit tests covering the new invoice line item helpers:

`priceId` and price retrieval
`unitAmount` handling for both `price_details` and `inline_price_data`
`taxBehavior` method

All test suites pass locally. Please refer to the CI checks for confirmation.

### Motivation
Taylor mentioned that the team “would love some assistance making Cashier compatible with Stripe's latest API version (Basil)” — this PR answers that call. It updates core invoice item logic and tests to ensure developers can rely on Cashier when using the latest Stripe features.

This is my first significant open source contribution attempt so apologies if it isn't quite up to scratch, I welcome any criticism or commits to improve it!

Diddyy committed:
Update branch alias and Stripe SDK

Diddyy committed:
Update Cashier.php to version 16.0.0 and change Stripe API version to 2025-03-31.basil

Diddyy committed:
Upgrade Stripe SDK to ^17.3.0, and update to work with latest API basil

Diddyy committed:
Refactor payments method in Invoice.php to retrieve payments via API, enhancing flexibility for expansion and filtering.

Diddyy committed:
Refactor discount, tax, and invoice line item retrieval in Invoice.php for improved clarity and performance; add static method to get Stripe SDK client in SubscriptionBuilder.php.

Diddyy committed:
Refactor invoiceLineItems method in Invoice.php to improve clarity; now accepts parameters for enhanced flexibility in retrieving Stripe line items.

Diddyy committed:
Update PHP version requirement in composer.json from ^8.2 to ^8.1 for compatibility adjustments.

Diddyy committed:
Ensure consistent spacing and formatting to pass StyleCI

Diddyy committed:
Further StyleCI fixes

Diddyy committed:
Add balance type and checkout session methods to CustomerBalanceTransaction; include unit tests for new functionality

driesvints commented:
Thank you. I'll review this one when I find some time.

kyranb commented on src/SubscriptionItem.php:
```diff
@@ -207,16 +208,37 @@ public function swapAndInvoice($price, array $options = [])
+        }
+
+        // Get the meter to get the event name
+        $meter = $this->subscription->owner->stripe()->billing->meters->retrieve($stripePrice->recurring->meter);
```
Just wondering if the event name should be stored locally?

Otherwise for an application that is reporting a lot of usage, the Stripe API might be hit a lot more than is necessary?

Diddyy replied:
Great idea, I have added that suggestion.

kyranb commented:
@driesvints 

Is there a rough target release date for Cashier 16/Basil support?

Diddyy committed:
Add meter_event_name support

Diddyy committed:
Merge pull request #1 (add-meter_event_name-to-subscription_items)

Add meter_event_name storage for metered items

Diddyy committed:
Merge remote-tracking branch 'origin/16.x' into 16.x

Diddyy committed:
refactor(invoice): include tax rate details & simplify retrieval

- Add new expansion paths so invoice responses include full tax_rate objects
- Simplify tax rate lookup to return the expanded object and eliminate extra Stripe API calls

Diddyy committed:
feat(subscription_items): add meter_id column and cache meter details

- Introduce nullable `meter_id` column in `subscription_items` (lines 12–17 in migration)
- Cast `meter_id` and `meter_event_name` as strings on the model for direct access
- Cache meter ID and event name on subscription swaps and price additions
- Enhance usage reporting to read from cache or fetch-and-store meter details on first call
- Add tests to ensure `meter_id` and `meter_event_name` persist after creation

Diddyy committed:
fix(styleCI): Fix space at end of file

Diddyy committed:
refactor(Invoice): Expand price objects when fetching invoice line items

Diddyy committed:
refactor: Use UUIDs for meter event identifiers

- uniqid can collide in high‑load situations. Laravel’s Str::uuid() provides truly unique identifiers.

Diddyy commented:
I feel like its been draft for too long. I have done all I personally can on it so any further suggestions I will implement but apart from that I feel like it would be ace to get the ball rolling on this being merged! @driesvints

Diddyy committed:
refactor(subscription_builder): validate empty price list in quantity()

- Add guard in `SubscriptionBuilder::quantity` to throw `InvalidArgumentException` if no prices are provided
- Introduce unit test to verify `quantity()` without any prices raises the expected exception

kyranb commented on src/Cashier.php:
```diff
@@ -19,14 +19,14 @@ class Cashier
      * @var string
      */
-    const STRIPE_VERSION = '2023-10-16';
+    const STRIPE_VERSION = '2025-03-31.basil';
```
When this is eventually merged and a new cashier version is tagged, it might be best to target the latest version of basil.

The latest basil version is now `2025-06-30.basil`

https://docs.stripe.com/changelog/basil#2025-06-30.basil

Diddyy replied:
It moves quick, I targeted the latest one when I first did this! But definitely something the Laravel Team can edit before merging. I feel like If i keep updating it now I will be chasing my tail as I have no idea when they'll merge it!

crynobone replied:
Shouldn't we rely on https://github.com/stripe/stripe-php/blob/v17.4.0/API_VERSION value?

crynobone replied:
Also, what will happen when `stripe/stripe-php` release a minor version using an updated API version? Do we need to lock the dependencies to patch release only?

kyranb replied:
Few relevant bits of information from Stripe:

_Basil is the second release in our [new API versioning model](http://stripe.com/blog/introducing-stripes-new-api-release-process). As the first version in the Basil release, 2025-03-31.basil introduces both breaking changes and new features—subsequent Basil versions will include only additive changes. (Our last release was [Acacia](https://docs.stripe.com/changelog/acacia): every time we introduce a breaking change version, we move to the next release.)_

https://docs.stripe.com/changelog/basil


_Every release includes a new SDK for each supported language and up-to-date documentation describing the latest features and how to upgrade. While Stripe’s SDKs for each language will continue to use semantic versioning, each SDK version is directly associated with an API release, ensuring that you can effectively manage updates._

https://stripe.com/blog/introducing-stripes-new-api-release-process

crynobone committed:
setup workbench

Signed-off-by: Mior Muhammad Zaki <crynobone@gmail.com>

crynobone committed:
chore: StyleCI Improvements

Signed-off-by: Mior Muhammad Zaki <crynobone@gmail.com>

crynobone committed:
wip

Signed-off-by: Mior Muhammad Zaki <crynobone@gmail.com>

crynobone committed:
update testbench.yaml and tests

Signed-off-by: Mior Muhammad Zaki <crynobone@gmail.com>

crynobone commented on src/Concerns/ManagesCustomer.php:
```diff
@@ -261,47 +263,161 @@ public function syncStripeCustomerDetails()
+     * By default, applies to the primary subscription only for safety.
      *
      * @param  string  $coupon
+     * @param  string|array|null  $subscriptionTypes  Specific subscription types to target, or null for primary only
```
Could you provide the API link and specify the supported values for the `array` parameter?

crynobone commented on src/Invoice.php:
```diff
@@ -553,13 +560,20 @@ public function tabPrice($price, $quantity = 1, array $options = [])
-     * Refresh the invoice.
+     * Refresh the invoice instance from Stripe.
      *
+     * @param  array  $expand
```
What's the possible type of `$expand`?

kyranb replied:
@crynobone 

See here:

https://docs.stripe.com/api/expanding_objects?api-version=2025-06-30.basil

https://docs.stripe.com/expand

crynobone replied:
```suggestion
     * @param  array<int, string>  $expand
```

crynobone commented on src/SubscriptionItem.php:
```diff
@@ -228,11 +279,92 @@ public function reportUsage($quantity = 1, $timestamp = null)
+            return null;
+        }
+
+        $date = Carbon::createFromTimestamp($stripeItem->current_period_start);
```
@driesvints Do you have any opinion whether to use `Carbon`, `CarbonImmutable` or use `Illuminate\Support\Date` so application developers can decide whether to use immutable or not?

driesvints replied:
No hard opinion but I'd say continue following what was used already. Personally I'd use `CarbonImmutable` but I don't know of the implications on the package.

crynobone commented on src/SubscriptionItem.php:
```diff
@@ -164,7 +179,9 @@ public function swap($price, array $options = [])
         $this->fill([
             'stripe_product' => $stripeSubscriptionItem->price->product,
             'stripe_price' => $stripeSubscriptionItem->price->id,
+            'meter_id' => $meterId,
             'quantity' => $stripeSubscriptionItem->quantity,
+            'meter_event_name' => $meterEventName,
```
```suggestion
            'quantity' => $stripeSubscriptionItem->quantity,
            'meter_id' => $meterId,
            'meter_event_name' => $meterEventName,
```

crynobone commented on src/SubscriptionBuilder.php:
```diff
@@ -208,6 +231,19 @@ public function anchorBillingCycleOn($date)
+    /**
+     * Set billing thresholds for the subscription.
+     *
+     * @param  array  $thresholds
```
what's the possible values of `$thresholds`?

kyranb replied:
https://docs.stripe.com/api/subscriptions/object?api-version=2025-06-30.basil#subscription_object-billing_thresholds

crynobone replied:
```suggestion
     * @param  array{amount_gte?: int|null, reset_billing_cycle_anchor?: bool|null}  $thresholds
```

crynobone commented on src/Subscription.php:
```diff
@@ -1301,24 +1414,72 @@ public function latestPayment()
+    /**
+     * Get all discounts that apply to the subscription.
+     *
+     * @return \Illuminate\Support\Collection|\Laravel\Cashier\Discount[]
```
Shouldn't this be

```suggestion
     * @return \Illuminate\Support\Collection<int, \Laravel\Cashier\Discount>
```

crynobone commented on src/Concerns/ManagesCustomer.php:
```diff
@@ -261,47 +263,161 @@ public function syncStripeCustomerDetails()
-     * Apply a coupon to the customer.
+     * Get all discounts that apply to the customer's subscriptions.
+     *
+     * @return \Illuminate\Support\Collection|\Laravel\Cashier\Discount[]
```
Shouldn't it be

```suggestion
     * @return \Illuminate\Support\Collection<int, \Laravel\Cashier\Discount>
```

crynobone commented on src/Concerns/ManagesCustomer.php:
```diff
@@ -261,47 +263,161 @@ public function syncStripeCustomerDetails()
+     * @throws \Laravel\Cashier\Exceptions\InvalidCoupon
      */
-    public function applyCoupon($coupon)
+    public function applyCoupon($coupon, $subscriptionTypes = null)
```
some method use `$coupon` while other use `$couponId`, might be best to standardise it.

crynobone commented on src/Coupon.php:
```diff
@@ -79,6 +79,26 @@ public function rawAmountOff()
+     *
+     * @return bool
+     */
+    public function isForeverAmountOff()
```
Maybe `isAmountOffWithoutDuration()`? 

Suggestion @driesvints @joetannenbaum

driesvints replied:
Just `isForever`?

crynobone commented on src/Invoice.php:
```diff
@@ -739,6 +771,164 @@ public function isPaid()
+    /**
+     * Get the invoice payments.
+     *
+     * @return \Illuminate\Support\Collection|\Laravel\Cashier\InvoicePayment[]
```
Shouldn't it be:

```suggestion
     * @return \Illuminate\Support\Collection<int, \Laravel\Cashier\InvoicePayment>
```

crynobone commented on src/Invoice.php:
```diff
@@ -739,6 +771,164 @@ public function isPaid()
+    /**
+     * Get the parent information for this invoice.
+     *
+     * @return object|null
```
Can we determine the class or it's just `object`?

crynobone commented on src/Invoice.php:
```diff
@@ -739,6 +771,164 @@ public function isPaid()
+    /**
+     * Get the subscription details for this invoice.
+     *
+     * @return object|null
```
Can we determine the class or it's just `object`?

crynobone commented on src/Invoice.php:
```diff
@@ -739,6 +771,164 @@ public function isPaid()
+     * Apply a coupon to this invoice.
+     *
+     * @param  string  $couponId
+     * @param  array  $options
```
What's the available values of `$options`?

crynobone commented on src/Invoice.php:
```diff
@@ -739,6 +771,164 @@ public function isPaid()
+    /**
+     * Apply a discount to this invoice.
+     *
+     * @param  array  $discount
```
What's the possible values of `$discount`?

crynobone committed:
Update src/SubscriptionItem.php

Diddyy committed:
Update src/SubscriptionBuilder.php

Co-authored-by: Mior Muhammad Zaki <crynobone@gmail.com>

Diddyy committed:
Update src/Invoice.php

Co-authored-by: Mior Muhammad Zaki <crynobone@gmail.com>

Diddyy committed:
Update src/Subscription.php

Co-authored-by: Mior Muhammad Zaki <crynobone@gmail.com>

Diddyy committed:
Update src/Invoice.php

Co-authored-by: Mior Muhammad Zaki <crynobone@gmail.com>

Diddyy committed:
Update src/Concerns/ManagesCustomer.php

Co-authored-by: Mior Muhammad Zaki <crynobone@gmail.com>

crynobone commented on src/Invoice.php:
```diff
@@ -480,42 +498,31 @@ public function invoiceItems()
-     * Get all of the invoice items.
+     * Get all of the invoice line items.
      *
+     * @param  array  $params
```
```suggestion
     * @param  array{ending_before?: string, expand?: string[], limit?: int, starting_after?: string}  $params
```

https://github.com/stripe/stripe-php/blob/0c65d3f93e0332be9a2e7486c16941e6a177cf70/lib/Service/InvoiceService.php#L48-L65

crynobone committed:
early return

Signed-off-by: Mior Muhammad Zaki <crynobone@gmail.com>

crynobone committed:
wip

Signed-off-by: Mior Muhammad Zaki <crynobone@gmail.com>

crynobone committed:
wip

Signed-off-by: Mior Muhammad Zaki <crynobone@gmail.com>

crynobone committed:
wip

Signed-off-by: Mior Muhammad Zaki <crynobone@gmail.com>

crynobone commented on src/Invoice.php:
```diff
@@ -609,15 +641,29 @@ protected function formatAmount($amount)
+     */
+    protected function getTaxRate($taxRateDetails)
+    {
+        // If tax_rate is already expanded as an object, return it
+        if (isset($taxRateDetails->tax_rate) && is_object($taxRateDetails->tax_rate) && isset($taxRateDetails->tax_rate->id)) {
+            return $taxRateDetails->tax_rate;
+        }
+
+        return null;
```
Shouldn't this check if `$taxRateDetails->tax_rate` is an instance of `Stripe\TaxRate`?

```suggestion
        return $taxRateDetails->tax_rate instanceof \Stripe\TaxRate ? $taxRateDetails->tax_rate : null;
```

crynobone committed:
wip

Signed-off-by: Mior Muhammad Zaki <crynobone@gmail.com>

crynobone committed:
wip

Signed-off-by: Mior Muhammad Zaki <crynobone@gmail.com>

crynobone committed:
wip

Signed-off-by: Mior Muhammad Zaki <crynobone@gmail.com>

crynobone committed:
wip

Signed-off-by: Mior Muhammad Zaki <crynobone@gmail.com>

crynobone committed:
wip

Signed-off-by: Mior Muhammad Zaki <crynobone@gmail.com>

crynobone committed:
wip

Signed-off-by: Mior Muhammad Zaki <crynobone@gmail.com>

crynobone committed:
wip

Signed-off-by: Mior Muhammad Zaki <crynobone@gmail.com>

crynobone committed:
wip

Signed-off-by: Mior Muhammad Zaki <crynobone@gmail.com>

crynobone committed:
wip

Signed-off-by: Mior Muhammad Zaki <crynobone@gmail.com>

crynobone committed:
wip

Signed-off-by: Mior Muhammad Zaki <crynobone@gmail.com>

crynobone committed:
wip

Signed-off-by: Mior Muhammad Zaki <crynobone@gmail.com>

crynobone committed:
wip

Signed-off-by: Mior Muhammad Zaki <crynobone@gmail.com>

crynobone committed:
wip

Signed-off-by: Mior Muhammad Zaki <crynobone@gmail.com>

crynobone committed:
wip

Signed-off-by: Mior Muhammad Zaki <crynobone@gmail.com>

crynobone committed:
wip

Signed-off-by: Mior Muhammad Zaki <crynobone@gmail.com>

crynobone committed:
wip

Signed-off-by: Mior Muhammad Zaki <crynobone@gmail.com>

crynobone committed:
Use `Model::asDateTime()`

Signed-off-by: Mior Muhammad Zaki <crynobone@gmail.com>

crynobone committed:
wip

Signed-off-by: Mior Muhammad Zaki <crynobone@gmail.com>

crynobone committed:
Use constant from Stripe

Signed-off-by: Mior Muhammad Zaki <crynobone@gmail.com>

crynobone committed:
Merge branch '16.x' into 16.x

taylorotwell commented:
@crynobone could you do a bullet point list upgrade guide of what users the steps users will need to take to upgrade?

crynobone committed:
Update src/Invoice.php

crynobone committed:
Update Invoice.php

crynobone commented on src/Concerns/ManagesCustomer.php:
```diff
@@ -263,47 +278,164 @@ public function syncStripeCustomerDetails()
+     */
+    public function applyCouponToAllSubscriptions($coupon)
+    {
+        return $this->applyCoupon($coupon, '*');
```
```suggestion
     * @param  string  $couponId
     * @return void
     *
     * @throws \Laravel\Cashier\Exceptions\InvalidCoupon
     */
    public function applyCouponToAllSubscriptions($couponId)
    {
        return $this->applyCoupon($couponId, '*');
```

crynobone committed:
Update src/Concerns/ManagesCustomer.php

crynobone commented on src/Invoice.php:
```diff
@@ -739,6 +781,172 @@ public function isPaid()
+    /**
+     * Apply a discount to this invoice.
+     *
+     * @param  array<string, mixed>  $discount
+     * @return $this
+     *
+     * @see https://docs.stripe.com/api/invoices/update#update_invoice-discounts
+     */
+    public function applyDiscount(array $discount)
+    {
+        /** @var \Stripe\Service\InvoiceService $invoiceService */
+        $invoiceService = $this->owner->stripe()->invoices;
+
+        $this->invoice = $invoiceService->update(
+            $this->invoice->id,
+            ['discounts' => [$discount]]
+        );
```
```suggestion
     * @param  array<string, mixed>  $discounts
     * @return $this
     *
     * @see https://docs.stripe.com/api/invoices/update#update_invoice-discounts
     */
    public function applyDiscount(array $discounts)
    {
        /** @var \Stripe\Service\InvoiceService $invoiceService */
        $invoiceService = $this->owner->stripe()->invoices;

        $this->invoice = $invoiceService->update(
            $this->invoice->id,
            ['discounts' => [$discounts]]
        );
```

crynobone committed:
Update src/Invoice.php

Diddyy commented:
@crynobone @taylorotwell @driesvints Whats currently holding this back? Keen to see it merged :)

taylorotwell commented on src/Concerns/ManagesInvoices.php:
```diff
@@ -200,15 +218,40 @@ public function upcomingInvoice(array $options = [])
+               isset($parameters['subscription_details']) ||
+               isset($parameters['schedule']) ||
+               isset($parameters['schedule_details']) ||
+               isset($parameters['invoice_items']);
```
Is this supposed to be `&&`?

taylorotwell commented:
@Diddyy merged, but I have no idea how to update the docs for this, and also need an upgrade guide.

kyranb commented:
Thanks @taylorotwell for merging.

One other thing to flag perhaps before this gets a tagged release. It seems that as part of the basil release, Stripe are moving towards a new "flexible" billing mode.

https://docs.stripe.com/billing/subscriptions/billing-mode
https://docs.stripe.com/changelog/basil/2025-06-30/billing-mode
https://docs.stripe.com/billing/subscriptions/billing-mode?dashboard-or-api=api&lang=php
https://docs.stripe.com/api/subscriptions/create#create_subscription-billing_mode

While it's possible to pass in the `'billing_mode'` when creating a new subscription/checkout like so

```
$user->newSubscription('default', $priceId)
      ->create(null, [], [
          'billing_mode' => ['type' => 'flexible']
      ]);
```

I'm wondering if it's worth having a config option and/or a helper method on the SubscriptionBuilder

i.e

```php
'cashier' => [
          'default_billing_mode' => env('CASHIER_DEFAULT_BILLING_MODE', 'classic'),
      ],
```

or 

```php
  class SubscriptionBuilder
  {
      /**
       * The billing mode for the subscription.
       *
       * @var array|null
       */
      protected $billingMode = null;

      /**
       * Set the billing mode for the subscription.
       *
       * @param  string  $type
       * @return $this
       */
      public function withBillingMode(string $type = 'flexible')
      {
          $this->billingMode = ['type' => $type];
          return $this;
      }

      /**
       * Get the default billing mode from config.
       *
       * @return string
       */
      protected function getDefaultBillingMode(): string
      {
          return config('cashier.default_billing_mode', 'classic');
      }

      /**
       * Get the effective billing mode for this subscription.
       *
       * @return string
       */
      protected function getEffectiveBillingMode(): string
      {
          if ($this->billingMode) {
              return $this->billingMode['type'];
          }

          return $this->getDefaultBillingMode();
      }

      /**
       * Get the billing mode for the Stripe payload.
       *
       * @return array|null
       */
      protected function getBillingModeForPayload()
      {
          $mode = $this->getEffectiveBillingMode();

          // Only include billing_mode if it's not classic (Stripe's default)
          if ($mode !== 'classic') {
              return ['type' => $mode];
          }

          return null;
      }
      
```

What do you think? @taylorotwell @driesvints  @crynobone

Diddyy commented:
> @Diddyy merged, but I have no idea how to update the docs for this, and also need an upgrade guide.

@crynobone is this something you're able to help with? :)

crynobone commented:
@Diddyy I'm going through the upgrading guide, and we did face some issues, especially dealing with `v2` specific API on an old Stripe account (old API keys and API versions).

<img width="1484" height="1007" alt="image" src="https://github.com/user-attachments/assets/974e5bf3-31e6-4e1b-bf10-cc6e4e1ab0df" />
<img width="483" height="252" alt="image" src="https://github.com/user-attachments/assets/73164c92-d82d-4bbd-96d6-cf0d2d029929" />

https://docs.stripe.com/api-v2-overview?api-version=2025-07-30.preview&rds=1#limitations

We need to be able to list all the required steps to upgrade to the new version.

crynobone commented:
Furthermore, new metered billings is not on the PR description.

Diddyy commented:
> @Diddyy I'm going through the upgrading guide, and we did face some issues, especially dealing with `v2` specific API on an old Stripe account (old API keys and API versions).
> 
> <img alt="image" width="1484" height="1007" src="https://private-user-images.githubusercontent.com/172966/479262775-974e5bf3-31e6-4e1b-bf10-cc6e4e1ab0df.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTU1NjA0OTQsIm5iZiI6MTc1NTU2MDE5NCwicGF0aCI6Ii8xNzI5NjYvNDc5MjYyNzc1LTk3NGU1YmYzLTMxZTYtNGUxYi1iZjEwLWNjNmU0ZTFhYjBkZi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwODE4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDgxOFQyMzM2MzRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hMTcyYzY5YjAxZTY2YjU1ZDlhNjJlMmM3YmFkZTNhZTFiZmI1OGZmYzM3ZjhlNDg3YWYzMjZiOTgyNjlkN2VhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.NdJkPsqqD5Zjc-gnu_fPESbHfYPasHayKB61Y4ywTxc"> <img alt="image" width="483" height="252" src="https://private-user-images.githubusercontent.com/172966/479262796-73164c92-d82d-4bbd-96d6-cf0d2d029929.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTU1NjA0OTQsIm5iZiI6MTc1NTU2MDE5NCwicGF0aCI6Ii8xNzI5NjYvNDc5MjYyNzk2LTczMTY0YzkyLWQ4MmQtNGJiZC05NmQ2LWNmMGQyZDAyOTkyOS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwODE4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDgxOFQyMzM2MzRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xYjExODEyNDljMDlhNWUzNWU5YWMzMWYxNDM1NzFmMjA1MTcyYmM4MTYyODQ3ZjVlY2U0NWZjNzAxMTVkMjdlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.SwFFt6abA-ZoOaZPeTymfvZv_95l43dQO-1hr5HN1SQ">
> https://docs.stripe.com/api-v2-overview?api-version=2025-07-30.preview&rds=1#limitations
> 
> We need to be able to list all the required steps to upgrade to the new version.

Thanks for checking this @crynobone 🙌

I won’t be able to put together the full upgrade guide myself, but I agree it’s needed to document the steps (API version upgrade, sandbox key requirements, metered billing, etc.). Maybe it makes sense to open a follow-up issue in the relevant repo dedicated to the upgrade guide so we can track that separately?
